### PR TITLE
[Testcase Refactoring] Decouple hardware coupling and add hw_classification in test_zero_redundancy_optimizer

### PR DESCRIPTION
**Pull request by zjq-86**
## Summary
This PR decouples hardware-specific assumptions in test_zero_redundancy_optimizer.py so the test suite runs on any accelerator supported by torch.accelerator, and adds hw_classification labels to all test classes for hardware-based test filtering. The PR also restores device-count protection that was lost when @skip_if_no_gpu was removed during class splitting.

### Hardware coupling fix: generalize hardcoded CUDA device indices
In `_test_zero_model_parallel`, change `torch.device(2 * self.rank)` / `torch.device(2 * self.rank + 1)` to `torch.device(device, 2 * self.rank)` / `torch.device(device, 2 * self.rank + 1)` so the device type is derived from `device` (which resolves to the current accelerator type via `torch.accelerator`) rather than defaulting to CUDA. 

### Device-count guard: add `_skip_if_insufficient_devices` runtime check
Add a `_skip_if_insufficient_devices(self, device_type)` helper method to `TestZeroRedundancyOptimizerDistributedBase` that checks `device_count() < self.world_size` at test execution time and calls `self.skipTest()` if insufficient devices are available. This restores the device-count protection previously provided by `@skip_if_no_gpu` (which checked both accelerator availability and device count), which was removed when the original `TestZeroRedundancyOptimizerDistributed` class was split into CPU-compatible and device-parametrized subclasses. The helper is called from all 8 device-parametrized test methods in `TestZeroRedundancyOptimizerDistributedDevice`:
- `test_step`
- `test_step_with_closure`
- `test_lr_scheduler`
- `test_multiple_param_groups`
- `test_collect_shards`
- `test_local_optimizer_parity`
- `test_zero_join_gpu`
- `test_ddp_zero_overlap`

### hw_classification labels
Add `hw_classification` class attributes to all test classes:
- `TestZeroRedundancyOptimizerSingleRank`: `HardwareClassification.GENERIC` — single-process (world_size=1), tests ZeRO constructor/state_dict/step logic without `instantiate_device_type_tests`.
- `TestZeroRedundancyOptimizerDistributed`: `HardwareClassification.GENERIC` — multi-process via `instantiate_parametrized_tests`, tests ZeRO partitioning algorithm (sorted-greedy) which is device-agnostic logic; uses `self.device` via `torch.accelerator` API with `@skip_if_no_gpu` gating but does not use `instantiate_device_type_tests`.
- `TestZeroRedundancyOptimizerDistributedDevice`: `HardwareClassification.ACCELERATOR` — device-parametrized via `instantiate_device_type_tests(TestZeroRedundancyOptimizerDistributedDevice, globals(), except_for="cpu")`, test methods accept `device` parameter.

## Motivation
The original test file hardcoded CUDA device mixed single-process, multi-process, and device-parametrized execution mechanisms in a single class, making it impossible to assign accurate hw_classification labels or run on non-CUDA accelerators. the tests can now run on any accelerator supported by PyTorch and be selected by hardware capability.

## Test Plan
Verified on CPU，NPU and GPU:
python test/distributed/optim/test_zero_redundancy_optimizer.py

## Notes
- `@skipIfHpu` on `test_zero_model_parallel` is retained for HPU-specific issues unrelated to hardware coupling.
- `except_for="cpu"` is kept in `instantiate_device_type_tests` for `TestZeroRedundancyOptimizerDistributedDevice` because these tests require an accelerator backend for distributed collectives.
- The test file consumes `torch.accelerator` API and `torch.get_device_module` directly; no framework constant is modified.
- depend on PR 197063（https://github.com/pytorch/pytorch/pull/197063）

### CODE REVIEW

File: test/distributed/optim/test_zero_redundancy_optimizer.py
Comment (JamesD18) [inline test/distributed/optim/test_zero_redundancy_optimizer.py:58]: Suggest removing it; global `device_type` must not be used.

File: test/distributed/optim/test_zero_redundancy_optimizer.py [hunk 1]
```diff
@@ -74,24 +71,45 @@ def deterministic_algorithms(enabled=True):
...
+        if dm.device_count() < self.world_size:
+            self.skipTest(
+                f"Skipping test: {dm.device_count()} {device_type} device(s) "
+                f"available, but world_size={self.world_size} requires "
+                f"{self.world_size} device(s)."
+            )
+
 
 class TestZeroRedundancyOptimizerSingleRank(TestZeroRedundancyOptimizer):
+    hw_classification = HardwareClassification.GENERIC
```
Comment (JamesD18) [inline test/distributed/optim/test_zero_redundancy_optimizer.py]: **GENERIC requires device-agnostic behavior**, but these tests resolve a device at runtime and pass it to create_pg, which initializes a real accelerator backend even at ws=1. Therefore, these tests should be categorized under ACCELERATOR.

File: test/distributed/optim/test_zero_redundancy_optimizer.py [hunk 2]
```diff
@@ -1083,12 +1121,13 @@ def join_process_group(self):
...
+    @onlyAccelerator
+    def test_zero_join_gpu(self, device):
         """Check that the ZeRO join hook allows training with uneven inputs
         on GPU."""
-        self._test_zero_join(self.device)
+        self._skip_if_insufficient_devices(self.device_type)
+        self._test_zero_join(self.device_type)
 
     @requires_gloo()
     def test_zero_join_cpu(self):
```
Comment (JamesD18) [inline test/distributed/optim/test_zero_redundancy_optimizer.py:1133]: Missing device parameters.Alternatively, could we consider splitting this out into a CPU-only test class?

File: test/distributed/optim/test_zero_redundancy_optimizer.py [hunk 3]
```diff
@@ -1392,24 +1396,42 @@ def test_ddp_zero_overlap(
...
         )
 
 
-instantiate_parametrized_tests(TestZeroRedundancyOptimizerSingleRank)
-instantiate_parametrized_tests(TestZeroRedundancyOptimizerDistributed)
+class TestZeroRedundancyOptimizerDistributedOnCPU(TestZeroRedundancyOptimizer):
+    hw_classification = HardwareClassification.CPU
+
+    @requires_gloo()
+    def test_zero_join_cpu(self):
```
Comment (JamesD18) [inline test/distributed/optim/test_zero_redundancy_optimizer.py]: missing device parameters

**Review by JamesD18 — APPROVED** [review]
LGTM.Thanks for addressing the review comments.
